pkg/codegen/python/gen.go

```diff
@@ -2990,16 +2969,245 @@ func GeneratePackage(tool string, pkg *schema.Package, extraFiles map[string][]b
 	}
 	files.Add(filepath.Join(pkgName, "pulumi-plugin.json"), plugin)
 
-	// Finally emit the package metadata (setup.py).
+	// Next, emit the package metadata (setup.py).
 	setup, err := genPackageMetadata(tool, pkg, pkgName, info.Requires, info.PythonRequires)
 	if err != nil {
 		return nil, err
 	}
 	files.Add("setup.py", []byte(setup))
 
+	// Finally, if pyproject.toml generation is enabled, generate
+	// this file and emit it as well.
+	if info.PyProject.Enabled {
+		project, err := genPyprojectTOML(
+			tool, pkg, pkgName,
+		)
+		if err != nil {
+			return nil, err
+		}
+		files.Add("pyproject.toml", []byte(project))
+	}
+
 	return files, nil
 }
 
+func genPyprojectTOML(tool string,
+	pkg *schema.Package,
+	pyPkgName string,
+) (string, error) {
+	// First, create a Writer for everything in pyproject.toml
+	w := &bytes.Buffer{}
+	// Create an empty toml manifest.
+	schema := new(PyprojectSchema)
+	schema.Project = new(Project)
+	// Populate the fields.
+	schema.Project.Name = &pyPkgName
+	schema.Project.Keywords = pkg.Keywords
+
+	// Setting dependencies fails if the deps we provide specify
+	// an invalid Pulumi package version as a dep.
+
+	err := setDependencies(schema, pkg)
+	if err != nil {
+		return "", err
+	}
+
+	// These values are hardcoded to Pulumi's contact information.
+	// TODO(@Robbie): Third-party provider users can override this field.
+	// Set "Authors" and "Maintainers" fields.
+	setContacts(schema)
+
+	// This sets the minimum version of Python.
+	setPythonRequires(schema, pkg)
+	// Set the project's URLs
+	setURLs(schema, pkg)
+
+	// These values are set the same way as in setup.py.
+	setLicense(schema, pkg)
+	setVersion(schema, pkg)
+	// Set the path to the README.
+	readme := "README.md"
+	schema.Project.README = &readme
+
+	// Marshal the data into TOML format.
+	err = toml.NewEncoder(w).Encode(schema)
+	return w.String(), err
+}
+
+// setVersion mutates the Pyproject schema to use the version found
+// in the provider schema, or defaulting to 0.0.0 which will be replaced
+// later.
+func setVersion(schema *PyprojectSchema, pkg *schema.Package) {
```

Comment:
This is very much a taste thing and I know it can be controversial but in general if a function isn't reused, isn't independently tested, and isn't doing anything very complex (which if it we're it ought to be independently tested as well!) I'd rather it just be inlined. Long functions are not hard to read, humans are pretty well adapted to reading long form text, we're not so good at jumping from reference to reference. There's a nice post about this from Carmack: http://number-none.com/blow/john_carmack_on_inlined_code.html

Reply:
I do agree that this is borderline overkill. Each of these methods is small enough that there's not much gained with the abstraction.

To get this through approval, it sounds like you're proposing two possible options:
1. just inline the code
2. add test cases for each of the functions

I'd personally prefer (2), because IMO the Pulumi codebase _does_ have too many long method, particularly in codegen, that make it hard to test, and it's better to risk rubberbanding to the other extreme to make sure this function doesn't become unmanageable overtime. 

> humans are pretty well adapted to reading long form text, we're not so good at jumping from reference to reference

I think tiny functions aren't optimizing for the case where you jump to them, but the case where you _don't_. 90% of the time someone is reading this function, they're not going to need to jump to `setVersion` because there's probably something else in this method that's actually causing the problem. So by abstracting the internals of `setVersion`, I'm removing those details from the mix when someone is tracing something of actual merit.

Nevertheless, I think I'll just go ahead and inline these tiny functions because you're right that it's at least questionably overkill. XD

Reply:
I inlined the smallest functions, but I left the ones that have actual logic as they were. I'll add a unit test for the pure function. The other function is already thoroughly tested through the codegen tests, since it calls into the same function as the traditional path.

Reply:
Actually, both paths are thoroughly tested already.